perception/detected_object_validation/src/obstacle_pointcloud_based_validator.cpp

```diff
@@ -85,24 +85,109 @@ ObstaclePointCloudBasedValidator::ObstaclePointCloudBasedValidator(
 }
+void ObstaclePointCloudBasedValidator::on3dObjectsAndObstaclePointCloud(
+  const autoware_auto_perception_msgs::msg::DetectedObjects::ConstSharedPtr & input_objects,
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & input_obstacle_pointcloud)
+{
+  autoware_auto_perception_msgs::msg::DetectedObjects output, removed_objects;
+  output.header = input_objects->header;
+  removed_objects.header = input_objects->header;
 
+  // Transform to pointcloud frame
+  autoware_auto_perception_msgs::msg::DetectedObjects transformed_objects;
+  if (!object_recognition_utils::transformObjects(
+        *input_objects, input_obstacle_pointcloud->header.frame_id, tf_buffer_,
+        transformed_objects)) {
+    return;
+  }
+
+  // Convert to PCL
+  pcl::PointCloud<pcl::PointXYZ>::Ptr obstacle_pointcloud(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::fromROSMsg(*input_obstacle_pointcloud, *obstacle_pointcloud);
+  if (obstacle_pointcloud->empty()) {
+    return;
+  }
+
+  // Create Kd-tree to search neighbor pointcloud to reduce cost
+  pcl::search::Search<pcl::PointXYZ>::Ptr kdtree =
+    pcl::make_shared<pcl::search::KdTree<pcl::PointXYZ>>(false);
+  kdtree->setInputCloud(obstacle_pointcloud);
+  for (size_t i = 0; i < transformed_objects.objects.size(); ++i) {
+    const auto & transformed_object = transformed_objects.objects.at(i);
+    const auto object_label_id = transformed_object.classification.front().label;
+    const auto & object = input_objects->objects.at(i);
+    const auto & transformed_object_position =
+      transformed_object.kinematics.pose_with_covariance.pose.position;
+    const auto search_radius = getMaxRadius3D(transformed_object);
+    if (!search_radius) {
+      output.objects.push_back(object);
+      continue;
+    }
+    // Search neighbor pointcloud to reduce cost
+    pcl::PointCloud<pcl::PointXYZ>::Ptr neighbor_pointcloud(new pcl::PointCloud<pcl::PointXYZ>);
+    std::vector<int> indices;
+    std::vector<float> distances;
+    pcl::PointXYZ trans_obj_position_pcl;
+    trans_obj_position_pcl.x = transformed_object_position.x;
+    trans_obj_position_pcl.y = transformed_object_position.y;
+    trans_obj_position_pcl.z = transformed_object_position.z;
+    kdtree->radiusSearch(trans_obj_position_pcl, search_radius.value(), indices, distances);
+    for (const auto & index : indices) {
+      neighbor_pointcloud->push_back(obstacle_pointcloud->at(index));
+    }
+
+    if (debugger_) debugger_->addNeighborPointcloud(neighbor_pointcloud);
+    // Filter object that have few pointcloud in them
+    const auto num = getPointCloudNumWithinShape(transformed_object, neighbor_pointcloud);
+    const auto object_distance =
+      std::hypot(transformed_object_position.x, transformed_object_position.y);
+    size_t min_pointcloud_num = std::clamp(
+      static_cast<size_t>(
+        points_num_threshold_param_.min_points_and_distance_ratio.at(object_label_id) /
+          object_distance +
```

Comment:
The formula here is correct in 2D, but incorrect in 3D.
I don't think it will be a big problem with the 2D formula since it is after height crop, but you should leave it in the comments just in case.

Reply:
@yukkysaito 
IMO:
- Since this is to adapt validation threshold as pointcloud density decreasing as far distance so the formular should be common for 2D and 3D. The min_points_num and max_points_num might be better changing for 2D and 3D. 
- The big difference happens when vehicle is in front of a up/down slope and the LiDAR resolution are largely changed in vertical axis. 
- As you mentioned, since we have height crop so I will leave it as is.

Reply:
@badai-nguyen I agree.
Could you please leave a light comment in the code? :pray: